src/Bicep.LangServer/BicepCompilationManager.cs

```diff
@@ -87,31 +91,66 @@ public void RefreshCompilation(DocumentUri documentUri)
                 {
                     UpsertCompilationInternal(documentUri, null, sourceFile);
                 }
+
                 return;
             }
 
             // TODO: This may cause race condition if the user is modifying the file at the same time
             // need to make a shallow copy so it counts as a different file even though all the content is identical
             // this was the easiest way to force the compilation to be regenerated
-            var shallowCopy = new BicepFile(compilationContext.Compilation.SourceFileGrouping.EntryPoint);
+            var shallowCopy = compilationContext.Compilation.SourceFileGrouping.EntryPoint.ShallowClone();
             UpsertCompilationInternal(documentUri, null, shallowCopy);
         }
 
-        public void UpsertCompilation(DocumentUri documentUri, int? version, string fileContents, string? languageId = null, bool triggeredByFileOpenEvent = false)
+        public void OpenCompilation(DocumentUri documentUri, int? version, string fileContents, string languageId)
         {
+            if (LanguageConstants.IsBicepOrParamsLanguage(languageId))
+            {
+                this.fileLanguageTracker.NotifyFileOpen(documentUri, languageId);
+            }
+
             if (this.ShouldUpsertCompilation(documentUri, languageId))
             {
-                var newFile = SourceFileFactory.CreateSourceFile(documentUri.ToUri(), fileContents);
-                UpsertCompilationInternal(documentUri, version, newFile, triggeredByFileOpenEvent: triggeredByFileOpenEvent);
+                var newFile = CreateSourceFile(documentUri, fileContents);
+                UpsertCompilationInternal(documentUri, version, newFile, triggeredByFileOpenEvent: true);
+            }
+        }
+
+        public void UpdateCompilation(DocumentUri documentUri, int? version, string fileContents)
+        {
+            if (this.ShouldUpsertCompilation(documentUri, languageId: null))
+            {
+                var newFile = CreateSourceFile(documentUri, fileContents);
+                UpsertCompilationInternal(documentUri, version, newFile, triggeredByFileOpenEvent: false);
             }
         }
 
+        private ISourceFile CreateSourceFile(DocumentUri documentUri, string fileContents)
+        {
+            // normally the language ID is only available on file open
+            // however we stashed the ID in a dictionary, so we can conveniently
+            // look it up for as long as the file is open
+            if (this.fileLanguageTracker.TryGetLanguageId(documentUri) is { } languageId)
+            {
+                // since we know what language we're dealing with, we can just create the file
+                return SourceFileFactory.CreateSourceFileByLanguageId(documentUri.ToUri(), fileContents, languageId);
+            }
+
+            if(SourceFileFactory.TryCreateArmTemplateFile(documentUri.ToUri(), fileContents) is { } sourceFile)
```

Comment:
We have well-known languageIds for json, jsonc & arm json:

https://github.com/Azure/bicep/blob/a70f5be863a3ae02930165e048e1e2bff3b2107e/src/Bicep.LangServer/Utils/DocumentSelectorFactory.cs#L22-L24

Reply:
I'm not convinced that we should try to rely on language IDs anymore. `SourceFileGroupingBuilder` is completely file extension based. (See my comment above.)